File: Astro-TFG/public/scripts/2-procesamiento-datos/prueba-1/version-2/python/main.py
import time
import tracemalloc
import gc
import js  # type: ignore
from pyscript import display
import numpy as np

seed = int(time.time() * 1000) % 2**32
rng = np.random.default_rng(seed)


def create_data_structure(size):
    return rng.integers(0, 1001, size=size)


def transform_data_structure(arr):
    return (arr**2 + np.log(arr + 1)) / np.sqrt(arr + 1)


def sort_data_structure(arr):
    return np.sort(arr)


def search_in_data_structure(arr, value):
    return np.any(arr == value)


def filter_data_structure(arr, threshold):
    return arr[arr > threshold]


def delete_from_data_structure(arr, value):
    return arr[arr != value]


def do_operations(size):
    metrics = {}
    start_total = time.perf_counter()
    tracemalloc.start()
    max_memory = 0

    start_op = time.perf_counter()
    my_array = create_data_structure(size)
    current_mem = tracemalloc.get_traced_memory()
    metrics['create'] = {
        'time': (time.perf_counter() - start_op) * 1000,
        'memory': current_mem[1] / (1024 * 1024)
    }
    max_memory = max(max_memory, current_mem[1] / (1024 * 1024))
    tracemalloc.stop()

    operations = [
        ('transform', transform_data_structure, (my_array,)),
        ('sort', sort_data_structure, (my_array,)),
        ('search', search_in_data_structure, (my_array, rng.choice(my_array))),
        ('filter', filter_data_structure, (my_array, 500)),
        ('delete', delete_from_data_structure, (my_array, rng.choice(my_array)))
    ]

    for op_name, op_func, args in operations:
        tracemalloc.start()
        start_op = time.perf_counter()
        gc.collect()

        result = op_func(*args)
        op_time = (time.perf_counter() - start_op) * 1000

        current_mem = tracemalloc.get_traced_memory()
        op_memory = current_mem[1] / (1024 * 1024)

        metrics[op_name] = {
            'time': op_time,
            'memory': abs(op_memory)
        }

        max_memory = max(max_memory, op_memory)
        tracemalloc.stop()

    metrics['output'] = {
        'total_time': (time.perf_counter() - start_total) * 1000,
        'memory_peak': max_memory
    }

    tracemalloc.stop()
    display_results(metrics)


def display_results(results):
    for op, data in results.items():
        if op != 'output':
            display(
                f"{op.upper()} - Time: {data['time']:.2f} ms | RAM: {data['memory']:.2f} MB",
                target="pyscript-output"
            )
    display(
        f"Total ET: {results['output']['total_time']:.2f} ms", target="pyscript-output"
    )
    display(
        f"RAM Peak: {results['output']['memory_peak']:.2f} MB", target="pyscript-output"
    )
    js.window.hideExecutionLoader()


def run_py_benchmark(event):
    js.clearCell('pyscript-output')
    do_operations(10_000_000)


js.window.run_py_benchmark = run_py_benchmark

File: Astro-TFG/public/scripts/2-procesamiento-datos/prueba-1/version-2/javascript/main.js
function create_data_structure(size) {
    return new Int32Array(size).map(() => Math.floor(Math.random() * 1001));
}

function transform_data_structure(arr) {
    const result = new Float64Array(arr.length);
    for (let i = 0; i < arr.length; i++) {
        const num = arr[i];
        result[i] = (num ** 2 + Math.log(num + 1)) / Math.sqrt(num + 1);
    }
    return result;
}

function sort_data_structure(arr) {
    const copy = new Float64Array(arr);
    return copy.sort((a, b) => a - b);
}

function search_in_data_structure(my_array, value) {
    return my_array.includes(value);
}

function filter_data_structure(my_array, threshold) {
    return my_array.filter(x => x > threshold);
}

function delete_from_data_structure(arr, value) {
    const count = arr.reduce((acc, val) => val !== value ? acc + 1 : acc, 0);
    const result = new arr.constructor(count);
    let index = 0;
    for (const val of arr) {
        if (val !== value) result[index++] = val;
    }
    return result;
}

let max_memory = 0;

function do_operations(size) {
    const metrics = {};
    const start_total = performance.now();

    const measureMemory = () => {
        if (performance.memory) {
            const current = performance.memory.usedJSHeapSize / (1024 * 1024);
            max_memory = Math.max(max_memory, current);
            return current;
        }
        return 0;
    };

    let start_op = performance.now();
    let my_array = create_data_structure(size);
    metrics['create'] = {
        time: performance.now() - start_op,
        memory: measureMemory(),
    };

    const operations = [
        ['transform', transform_data_structure, [my_array]],
        ['sort', sort_data_structure, [my_array]],
        [
            'search',
            search_in_data_structure,
            [my_array, my_array[Math.floor(Math.random() * my_array.length)]],
        ],
        ['filter', filter_data_structure, [my_array, 500]],
        [
            'delete',
            delete_from_data_structure,
            [my_array, my_array[Math.floor(Math.random() * my_array.length)]],
        ],
    ];

    for (const [op_name, op_func, args] of operations) {
        start_op = performance.now();
        if (window.gc) window.gc();
        const mem_before = performance.memory
            ? performance.memory.usedJSHeapSize / (1024 * 1024)
            : 0;
        let result = op_func(...args);

        const mem_after = performance.memory
            ? performance.memory.usedJSHeapSize / (1024 * 1024)
            : 0;
        metrics[op_name] = {
            time: performance.now() - start_op,
            memory: Math.abs(mem_after - mem_before),
        };
        max_memory = Math.max(max_memory, mem_after);
    }

    metrics['output'] = {
        total_time: performance.now() - start_total,
        memory_peak: max_memory,
    };

    displayResults(metrics);
}


function displayResults(results) {
    const output = document.getElementById("javascript-output");

    for (const [op, data] of Object.entries(results)) {
        if (op !== 'output') {
            const timeDiv = document.createElement("div");
            timeDiv.textContent =
                `${op.toUpperCase()} - Time: ${data.time.toFixed(2)} ms | RAM: ${data.memory.toFixed(2)} MB`;
            output.appendChild(timeDiv);
        }
    }

    const timeTotalDiv = document.createElement("div");
    timeTotalDiv.textContent =
        `Total ET: ${results.output.total_time.toFixed(2)} ms`;
    output.appendChild(timeTotalDiv);

    const memoryDiv = document.createElement("div");
    memoryDiv.textContent =
        `RAM Peak: ${results.output.memory_peak.toFixed(2)} MB`;
    output.appendChild(memoryDiv);
}


window.runJsBenchmark = function () {
    window.clearCell("javascript-output");
    window.showExecutionLoader();

    requestAnimationFrame(() => {
        setTimeout(() => {
            do_operations(10_000_000);
            window.hideExecutionLoader();
        }, 0);
    });
};